docs/cpu_templates/cpu-templates.md

```diff
@@ -0,0 +1,190 @@
+
+Firecracker supports two types of CPU templates:
+
+- Static CPU templates - a set of built-in CPU templates for users
+  to choose from
+- Custom CPU templates - users can create their own CPU templates in json
+  format and pass them to Firecracker
+
+> **Note**
+CPU templates for ARM (both static and custom) require the following patch
+to be available in the host kernel: [Support writable CPU ID registers from userspace](https://lore.kernel.org/kvm/20230212215830.2975485-1-jingzhangos@google.com/#t).
+Otherwise KVM will fail to write to the ARM registers.
+
+## Static CPU templates
+
+At the moment the following set of static CPU templates is supported:
+
+| CPU template | CPU vendor | CPU model             |
+|--------------|------------|-----------------------|
+| C3           | Intel      | any                   |
+| T2           | Intel      | any                   |
+| T2A          | AMD        | Milan                 |
+| T2CL         | Intel      | Cascade Lake or newer |
+| T2S          | Intel      | any                   |
+| V1N1         | ARM        | Neoverse V1           |
+
+T2 and C3 templates are mapped as close as possible to AWS T2 and C3 instances
+in terms of CPU features.
+
+The T2S template is designed to allow migrating [snapshots](../snapshotting/versioning.md)
+between hosts with Intel Skylake and Intel Cascade Lake securely by further
+restricting CPU features for the guest, however this comes with a performance
+penalty. Users are encouraged to carry out a performance assessment if they wish
+to use the T2S template.
+
+The T2CL template is mapped to be close to Intel Cascade Lake.
+It is not safe to use it on Intel CPUs older than Cascade Lake (such as Skylake).
+
+The only AMD template is T2A. It is considered safe to be used with AMD Milan.
+
+Intel T2CL and AMD T2A templates together aim to provide instruction set feature
+parity between CPUs running them, so they can form a heterogeneous fleet
+exposing the same instruction sets to the application.
+
+The V1N1 template is designed to represent ARM Neoverse V1 as ARM Neoverse N1.
+
+### Configuring static CPU templates
+
+Configuration of a static CPU template is performed via `/machine-config` API endpoint:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i  \
+  -X PUT 'http://localhost/machine-config' \
+  -H 'Accept: application/json'            \
+  -H 'Content-Type: application/json'      \
+  -d '{
+           "vcpu_count": 2,
+           "mem_size_mib": 1024,
+           "cpu_template": "T2CL"
+  }'
+```
+
+## Custom CPU templates
+
+Users can create their own CPU templates by creating a json file containing
+modifiers for CPUID, MSRs or ARM registers.
+
+> **Note**
+Creating custom CPU templates requires expert knowledge of
+CPU architectures. An inappropriate configuration may lead to guest crashes or
+making guests vulnerable to security attacks. Custom CPU templates must be
```

Comment:
This seems to conflict with the statement given earlier when it comes to the security affects of CPU templates.

This earlier statement being:

> CPU templates is not a security feature. It does not improve
the guest's security posture in any way. It is only supposed to modify